NXP support imx95 edma5 uart use case

yongxu-wang15 opened the pull request:
1. add edma v5 into dma_mcux_edma zephyr dma driver.
2. test pass with imx95 uart_async_api case with this edma5.
3. support imx95 cm7 into sdk-ng

yongxu-wang15 committed:
modules: hal_nxp: support imx95 cm7 into sdk ng

- support imx95 cm7 platform into sdk ng.

Signed-off-by: Yongxu Wang <yongxu.wang@nxp.com>

LaurentiuM1234 reviewed (commented):
Some DTS-related nitpicks below

LaurentiuM1234 commented:
Pls have a look at the twister/compliance failures

yongxu-wang15 committed:
boards: nxp: imx95_evk_mimx9596_m7: add uart dma support

- verify in uart_async_api test case.

Signed-off-by: Yongxu Wang <yongxu.wang@nxp.com>

yongxu-wang15 committed:
drivers: dma: Update NXP EDMA driver for version 5

1. The edma version 5 share one driver with edma 4.
2. Edma5 tcd structure some difference, Use tcd type to distinguish,
   and Edma5 uses 64 bytes for alignment instead of 32.
3. Some platforms have some address offsets for certain memory
   when processing from a DMA perspective, such as imx95 cm7 TCM,
   so add offset processing.

Signed-off-by: Yongxu Wang <yongxu.wang@nxp.com>

yongxu-wang15 commented on dts/arm/nxp/nxp_imx95_m7.dtsi:
```diff
@@ -143,6 +143,19 @@
 		};
 
+		edma5: dma-controller@42000000 {
+			#dma-cells = <2>;
```
thanks reminder, have updated this dts note

yongxu-wang15 commented on dts/arm/nxp/nxp_imx95_m7.dtsi:
```diff
@@ -234,6 +247,8 @@
 			reg = <0x42570000 DT_SIZE_K(64)>;
 			interrupts = <64 3>;
 			clocks = <&scmi_clk IMX95_CLK_LPUART3>;
+			dmas = < &edma5 0 529>, < &edma5 1 530>;
```
Thanks, I have deleted the extra space

yongxu-wang15 commented on dts/arm/nxp/nxp_imx95_m7.dtsi:
```diff
@@ -250,6 +265,8 @@
 			reg = <0x42590000 DT_SIZE_K(64)>;
 			interrupts = <66 3>;
 			clocks = <&scmi_clk IMX95_CLK_LPUART5>;
+			dmas = < &edma5 2 533>, < &edma5 3 534>;
```
sure, updated

yongxu-wang15 commented:
> Pls have a look at the twister/compliance failures

hi @LaurentiuM1234 , have fixed the twister and compliance failures now, please help to review it again

LaurentiuM1234 commented on modules/hal_nxp/mcux/CMakeLists.txt:
```diff
@@ -16,7 +16,6 @@ if(CONFIG_SOC_LPC11U66
     OR CONFIG_SOC_MIMX8UD7_ADSP
     OR CONFIG_SOC_MIMX9352_A55
     OR CONFIG_SOC_MIMX9596_A55
-    OR CONFIG_SOC_MIMX9596_M7
```
AFAIK this'll break the other EDMA driver which is why it was introduced via https://github.com/zephyrproject-rtos/zephyr/pull/87817. Is this really necessary?

CkovMk replied:
Indeed, the fsl_edma4 driver is used to support edma3, edma4 and edma5. I've been watching Zephyr edma driver for a while, and I would expect the same in Zephyr. @yongxu-wang15 could you extend the driver to support edma4 as well?

zejiang0jason replied:
Hi @LaurentiuM1234 , the fsl_edma_v2, and fsl_edma_soc_v2 can be added in sdk-ng, it will simplify the cases such as IMX95 M7 support. The only change is the contribution workflow, the patch need to be pushed to MCUX SDK repo, then synced to hal_nxp. thanks.

CkovMk commented on drivers/dma/dma_mcux_edma.c:
```diff
@@ -30,8 +30,14 @@ LOG_MODULE_REGISTER(dma_mcux_edma, CONFIG_DMA_LOG_LEVEL);
 #define HAS_CHANNEL_GAP(n)		DT_INST_NODE_HAS_PROP(n, channel_gap) ||
 #define DMA_MCUX_HAS_CHANNEL_GAP	(DT_INST_FOREACH_STATUS_OKAY(HAS_CHANNEL_GAP) 0)
 
+#if defined(CONFIG_DMA_MCUX_EDMA_V5)
```
The same SOC can have different version of EDMA instances. Say i.MX93 has EDMA3 and EDMA4, some may even have EDMA3,4,5 simutaniously. Such code may cause error.

yongxu-wang15 commented on dts/arm/nxp/nxp_imx95_m7.dtsi:
```diff
@@ -143,6 +143,19 @@
 			status = "disabled";
 		};
 
+		edma5: dma-controller@42000000 {
```
hi @LaurentiuM1234 , have updated the dts node, only reserved this edma2 node, please help review again

LaurentiuM1234 commented on dts/arm/nxp/nxp_imx95_m7.dtsi:
```diff
@@ -269,7 +278,7 @@
 			interrupts = <170 0>;
 			dai-index = <3>;
 			mclk-is-output;
-			dmas = <&edma2 30 60>, <&edma2 31 61>;
```
why change the channel indexes?

yongxu-wang15 replied:
this twister issue is produced by another PR: https://github.com/zephyrproject-rtos/zephyr/pull/87881/files
![image](https://github.com/user-attachments/assets/4413855e-ed5c-4b3e-b1ec-296af6a18ea5)
it can be fixed after developer merge patch into coummunity.
Hi @LaurentiuM1234 , This is modified by referring to other dts that use this driver. Sequential channel changes will make interrupts and interrupt sharing masks more readable, and imx95 have dma mux function, based on rightly mux request, it not impact expected function from my view

LaurentiuM1234 reviewed (dismissed):
Almost there, thanks. @yongxu-wang15 can you please take a look at the new twister failures?

Also, blocking this until https://github.com/zephyrproject-rtos/hal_nxp/pull/546 is sorted out to avoid any breakages.

yongxu-wang15 commented:
@LaurentiuM1234 have migrate his driver into mcu-sdk-ng now, please help review this PR again, thanks a lot
